File: services/CoinAPIService.py
import aiohttp
import os
from typing import Optional, Dict

class CoinAPIService:
    def __init__(self):
        self.api_key = os.getenv('COINAPI_KEY')
        if not self.api_key:
            raise ValueError("COINAPI_KEY environment variable not set")
        
        self.base_url = "https://rest.coinapi.io/v1"
        self.headers = {
            'X-CoinAPI-Key': self.api_key,
            'Accept': 'application/json'
        }
        
        # Cache for symbol mapping
        self._symbol_map = {
            'USDT': 'USDT',
            'WBNB': 'BNB',  # Map WBNB to BNB for CoinAPI
            'ETH': 'ETH',
            'BTC': 'BTC',
            'CAKE': 'CAKE',
            # Add more mappings as needed
        }
        
    async def get_token_price(self, symbol: str) -> Optional[Dict]:
        """
        Get token price from CoinAPI
        Returns: Dict with price and other metadata or None if not found
        """
        try:
            # Map token symbol to CoinAPI format
            api_symbol = self._symbol_map.get(symbol, symbol)
            
            # Use USD as quote currency
            endpoint = f"/exchangerate/{api_symbol}/USD"
            
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    f"{self.base_url}{endpoint}",
                    headers=self.headers
                ) as response:
                    if response.status == 200:
                        data = await response.json()
                        return {
                            'price': float(data['rate']),
                            'timestamp': data['time'],
                            'source': 'coinapi'
                        }
                    elif response.status == 429:
                        print("CoinAPI rate limit reached")
                        return None
                    else:
                        print(f"CoinAPI error: {response.status}")
                        return None
                        
        except Exception as e:
            print(f"Error fetching price from CoinAPI: {e}")
            return None
            
    async def get_exchange_rate(self, base_symbol: str, quote_symbol: str) -> Optional[float]:
        """
        Get exchange rate between two tokens
        Returns: Exchange rate or None if not found
        """
        try:
            base = self._symbol_map.get(base_symbol, base_symbol)
            quote = self._symbol_map.get(quote_symbol, quote_symbol)
            
            endpoint = f"/exchangerate/{base}/{quote}"
            
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    f"{self.base_url}{endpoint}",
                    headers=self.headers
                ) as response:
                    if response.status == 200:
                        data = await response.json()
                        return float(data['rate'])
                    return None
                    
        except Exception as e:
            print(f"Error fetching exchange rate from CoinAPI: {e}")
            return None

File: services/CoinAPIService.js
const axios = require('axios');

class CoinAPIService {
    constructor() {
        // Try multiple environment variable names
        this.apiKey = process.env.COINAPI_KEY || process.env.COIN_API_KEY || process.env.COIN_API_SECRET;
        
        // Check for API key
        if (!this.apiKey) {
            console.warn("COINAPI_KEY environment variable not set - price fetching will be limited");
            this.enabled = false;
            return;
        }
        
        this.enabled = true;
        this.baseUrl = "https://rest.coinapi.io/v1";
        this.headers = {
            'X-CoinAPI-Key': this.apiKey,
            'Accept': 'application/json'
        };
        
        // Symbol mapping for CoinAPI
        this.symbolMap = {
            'USDT': 'USDT',
            'WBNB': 'BNB',  // Map WBNB to BNB for CoinAPI
            'ETH': 'ETH',
            'WETH': 'ETH',  // Map WETH to ETH
            'BTC': 'BTC',
            'CAKE': 'CAKE',
            'LINK': 'LINK'
        };
    }
    
    async getTokenPrice(symbol) {
        if (!this.enabled) {
            return null;
        }

        try {
            // Map token symbol to CoinAPI format
            const apiSymbol = this.symbolMap[symbol] || symbol;
            
            // Add retry logic
            const maxRetries = 3;
            let retryCount = 0;
            let lastError = null;

            while (retryCount < maxRetries) {
                try {
                    const response = await axios.get(
                        `${this.baseUrl}/exchangerate/${apiSymbol}/USD`,
                        { 
                            headers: this.headers,
                            timeout: 5000 // 5 second timeout
                        }
                    );
                    
                    if (response.status === 200) {
                        return {
                            price: parseFloat(response.data.rate),
                            timestamp: response.data.time,
                            source: 'coinapi'
                        };
                    }
                    return null;
                } catch (error) {
                    lastError = error;
                    if (error.response?.status === 429) {
                        // Rate limit hit - wait longer between retries
                        await new Promise(resolve => setTimeout(resolve, 2000 * (retryCount + 1)));
                    }
                    retryCount++;
                }
            }

            // Log the final error after all retries
            if (lastError.response?.status === 429) {
                console.warn("CoinAPI rate limit reached after retries");
            } else {
                console.error(`Failed to fetch price from CoinAPI for ${symbol} after ${maxRetries} attempts:`, lastError.message);
            }
            return null;
            
        } catch (error) {
            if (error.response?.status === 429) {
                console.warn("CoinAPI rate limit reached");
            } else {
                console.error(`Error fetching price from CoinAPI for ${symbol}:`, error.message);
            }
            return null;
        }
    }
    
    async getExchangeRate(baseSymbol, quoteSymbol) {
        try {
            const base = this.symbolMap[baseSymbol] || baseSymbol;
            const quote = this.symbolMap[quoteSymbol] || quoteSymbol;
            
            const response = await axios.get(
                `${this.baseUrl}/exchangerate/${base}/${quote}`,
                { headers: this.headers }
            );
            
            if (response.status === 200) {
                return parseFloat(response.data.rate);
            }
            return null;
            
        } catch (error) {
            console.error(`Error fetching exchange rate from CoinAPI:`, error.message);
            return null;
        }
    }
}

module.exports = CoinAPIService;
